Source/options.h

```diff
@@ -0,0 +1,230 @@
+#include "SDL_stdinc.h"
+#include "../3rdParty/Storm/Source/storm.h"
+
+DEVILUTION_BEGIN_NAMESPACE
+
+/**
+ * @brief Base non-templated class for all options. Provides saving and loading functionality and a way to group every
+ * type of option together.
+*/
+class OptionBase {
+public:
+	/**
+	 * @brief Loads data from the ini into this instance.
+	 * @param sectionName The name of the section that contains the data for this option.
+	*/
+	virtual void Load(const std::string& sectionName) = 0;
+
+	/**
+	 * @brief Saves the data contained in this instance to the appropriate section in the ini.
+	 * @param sectionName The name of the section where the data from this instance will be saved to.
+	*/
+	virtual void Save(const std::string& sectionName) = 0;
+};
+
+/**
+ * @brief Base templated class for all options. Stores information about a given option, including its current value,
+ * default value and the name of the key inside the ini file that corresponds to it.
+ * @tparam T The underlying type of the setting.
+*/
+template <typename T>
+class Option : public OptionBase {
+public:
+	/**
+	 * @brief Initializes a new instance of the Option<T> with the specified display name and default value.
+	 * @param displayName The name to use when saving and loading the option to ini.
+	 * @param defaultValue The default value to use when the option does not yet exist.
+	*/
+	Option(const std::string& displayName, const T& defaultValue)
+	    : m_value(0)
+	    , m_displayName(displayName)
+	    , m_defaultValue(defaultValue)
+	{
+	}
+
+	/**
+	 * @brief Extracts the underlying value from this option instance.
+	 * @return The underlying value of the option.
+	*/
+	const T& operator*() const
+	{
+		return m_value;
+	}
+
+protected:
+	T m_value;
+	std::string m_displayName;
+	T m_defaultValue;
+};
+
+/**
+ * @brief A class that models a boolean option.
+*/
+class BooleanOption final : public Option<bool> {
+	using Option<bool>::Option;
+
+public:
+	void Load(const std::string& sectionName);
+	void Save(const std::string& sectionName);
+
+	/**
+	 * @brief Sets the underlying value of this option instance.
+	 * @param value The new value to be set as the current value of this option.
+	 * @return This instance with the new value set.
+	*/
+	BooleanOption& operator=(bool value);
+};
+
+/**
+ * @brief A class that models a group of options persisted in an ini section.
+*/
+class OptionGroup {
+public:
+	/**
+	 * @brief Initializes a new instance of the option group with the specified section name.
+	 * @param sectionName The name of the ini section used to read and write values.
+	*/
+	OptionGroup(const std::string& sectionName);
+
+	/**
+	 * @brief Loads the data from the ini file into this option group.
+	*/
+	void Load();
+
+	/**
+	 * @brief Saves the data stored in this option group to the ini file.
+	*/
+	void Save();
+
+protected:
+	void AddOption(OptionBase&);
+	void AddOptionGroup(OptionGroup&);
+
+private:
+	std::vector<OptionBase*> m_options;

```

Comment:
Did you run `clang-tidy` over this? It should've complained that the name does not follow the standard we've chosen.

Reply:
I don't know how to run that. Doesn't it run on the build?

I picked this format from the webkit page. I see no mention of member names in our clang-format file.

https://webkit.org/code-style-guidelines/
> Data members in C++ classes should be private. Static data members should be prefixed by “s_”. Other data members should be prefixed by “m_”.

Reply:
I've set up clang-tidy to follow the Google Style guide following our conversation in chat.

In the Google Style Guide this would be named `options_`.

Set up clang-tidy + Visual Studio integration and it should be able to automatically update all names to follow the configuration.

This is configured here: https://github.com/diasurgical/devilutionX/blob/7d926a4619a4400d1593056b9041e3bd99d05162/.clang-tidy#L61

Reply:
It does not run as part of CI yet as there would be too many warnings.
We probably shouldn't do a massive cleanup of the whole codebase until after 1.2.0 release but we should start running it on new code.